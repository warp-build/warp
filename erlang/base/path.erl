-module(path).

-export([add_extension/2]).
-export([contains/2]).
-export([dirname/1]).
-export([ensure/1]).
-export([extension/1]).
-export([filename/1]).
-export([from_list/1]).
-export([is_prefix/2]).
-export([join/2]).
-export([new/1]).
-export([nth_tail/2]).
-export([relativize/1]).
-export([strip_prefix/2]).
-export([tail/1]).
-export([to_list/1]).
-export([to_string/1]).

-export_type([t/0]).

-opaque t() :: binary().

new(Str) -> binary:list_to_bin([Str]).

-spec is_prefix(Prefix :: t(), Path :: t()) -> bool().
is_prefix(Prefix, Path) -> strip_prefix(Prefix, Path) =/= Path.

-spec nth_tail(Path :: t(), N :: non_neg_integer()) -> t().
nth_tail(Path, N) when N > 0 ->
  Parts = path:to_list(Path),
  path:from_list(lists:nthtail(length(Parts) - N, Parts)).

-spec strip_prefix(Prefix :: t(), Path :: t()) -> t().
strip_prefix(Prefix, Path) ->
  Replaced = str:new(string:replace(Path, Prefix, "")),
  Final = case str:begins_with(Replaced, "/") of
            true -> str:new(string:replace(Replaced, "/", ""));
            false -> Replaced
          end,
  new([ Final ]).

join(A, B) -> from_list([A, B]).

extension(Path) -> new(filename:extension(Path)).

add_extension(Path, Ext) -> new([Path, ".", Ext]).

filename(Path) -> new(filename:basename(Path)).

dirname(Path) -> new(filename:dirname(Path)).

contains(Path, Str) -> not (string:find(Path, Str) == nomatch).

to_list(Path) -> [ new(Seg) || Seg <- filename:split(Path) ].

to_string(Path) -> binary:bin_to_list(Path).

from_list([]) -> new("");
from_list(Ls) when is_list(Ls) -> new(filename:join(Ls)).

-spec relativize(Path :: t()) -> t().
relativize(Path) ->
  case to_list(Path) of
    [<<".">> | Rest] -> from_list(Rest);
    Parts -> from_list(Parts)
  end.

-spec tail(Path :: t()) -> t().
tail(Path) -> filename:join(tl(filename:split(Path))).

-spec ensure(Path :: t()) -> ok.
ensure(Path) -> filelib:ensure_dir(Path).
