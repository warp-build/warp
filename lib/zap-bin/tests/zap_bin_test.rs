use assert_cmd::Command;
use std::path::PathBuf;

fn run(parts: &[&str]) -> String {
    let root = std::fs::canonicalize(PathBuf::from(".")).unwrap();
    let parts = parts.to_vec();
    let mut cmd = Command::cargo_bin(parts[0]).unwrap();
    cmd.env("RUST_BACKTRACE", "1");
    cmd.env("ZAP_LOG", "info");
    cmd.env("ZAP_HOME", "../_zap_home");
    cmd.current_dir(std::fs::canonicalize(PathBuf::from("./tests/sample_project")).unwrap());
    cmd.arg("--quiet");
    for arg in &parts[1..] {
        cmd.arg(arg);
    }
    let bytes = cmd.assert().get_output().stdout.to_vec();

    std::env::set_current_dir(root).unwrap();

    String::from_utf8(bytes).unwrap()
}

#[test]
#[ignore]
pub fn zap_workspace_info() {
    insta::assert_snapshot!(run(&["zap", "workspace", "info"]))
}

#[test]
#[ignore]
pub fn zap_toolchains_list_active() {
    insta::assert_snapshot!(run(&["zap", "toolchains", "list-active"]))
}

#[test]
#[ignore]
pub fn zap_toolchains_list_archives() {
    insta::assert_snapshot!(run(&["zap", "toolchains", "list-archives"]))
}

#[test]
#[ignore]
pub fn zap_toolchains_list_available() {
    insta::assert_snapshot!(run(&["zap", "toolchains", "list-available"]))
}

#[test]
#[ignore]
pub fn zap_targets_list() {
    insta::assert_snapshot!(run(&["zap", "targets", "list"]))
}

#[test]
#[ignore]
pub fn zap_rules_list() {
    insta::assert_snapshot!(run(&["zap", "rules", "list"]))
}

#[test]
#[ignore]
pub fn zap_rules_dump_actions() {
    insta::assert_snapshot!(run(&["zap", "rules", "dump-actions", "//..."]))
}

#[test]
#[ignore]
pub fn zap_rules_dump_outputs() {
    insta::assert_snapshot!(run(&["zap", "rules", "dump-outputs", "//..."]))
}

#[test]
#[ignore]
pub fn zap_dep_graph_print() {
    // insta::assert_snapshot!(run(&["zap", "dep-graph", "print", "//..."]))
}

#[test]
#[ignore]
pub fn zap_cache_clear() {
    // insta::assert_snapshot!(run(&["zap", "cache", "clear", "//a:lib"]))
}

#[test]
#[ignore]
pub fn zap_build() {
    insta::assert_snapshot!(run(&["zap"]));
    insta::assert_snapshot!(run(&["zap", "build"]));
    insta::assert_snapshot!(run(&["zap", "build", "//..."]));
    insta::assert_snapshot!(run(&["zap", "build", "//a:lib"]));
}
