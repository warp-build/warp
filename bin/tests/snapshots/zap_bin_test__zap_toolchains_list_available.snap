---
source: lib/zap-bin/tests/zap_bin_test.rs
expression: "run(&[\"zap\", \"toolchains\", \"list-available\"])"
---
Available Toolchains: 
* :clojerl
* :deno
* :docker
* :elixir

